operator/src/main/java/oracle/kubernetes/operator/Main.java

```diff
@@ -447,94 +421,237 @@ ServiceWatcher getServiceWatcher(String namespace) {
+    Step readExistingNamespaces() {
+      return new CallBuilder()
+            .withLabelSelectors(selectionStrategy.getLabelSelectors())
+            .listNamespaceAsync(new NamespaceListStep());
+    }
+
+    private class NamespaceListStep extends DefaultResponseStep<V1NamespaceList> {
+
+      NamespaceListStep() {
+        super(new NamespaceListAfterStep());
       }
 
+      // If unable to list the namespaces, we may still be able to start them if we are using
+      // a strategy that specifies them explicitly.
       @Override
-      public String getLabelSelector() {
-        return tuningAndConfig().get("domainNamespaceLabelSelector");
+      protected NextAction onFailureNoRetry(Packet packet, CallResponse<V1NamespaceList> callResponse) {
+        return getConfiguredDomainNamespaces() != null && isNotAuthorizedOrForbidden(callResponse)
+                ? doNext(startNamespaces(Namespaces.this, getConfiguredDomainNamespaces()), packet) :
+                super.onFailureNoRetry(packet, callResponse);
       }
-    },
-    RegExp {
-      @Override
-      public boolean isRequireList() {
-        return true;
+
+      private Collection<String> getConfiguredDomainNamespaces() {
+        return selectionStrategy.getConfiguredDomainNamespaces();
       }
 
       @Override
-      public String getRegExp() {
-        return tuningAndConfig().get("domainNamespaceRegExp");
+      public NextAction onSuccess(Packet packet, CallResponse<V1NamespaceList> callResponse) {
+        final String intialResourceVersion = getResourceVersion(callResponse.getResult());
+        final Set<String> namespacesToStartNow = getNamespacesToStart(getNames(callResponse.getResult()));
+        getFoundDomainNamespaces(packet).addAll(namespacesToStartNow);
+
+        return doContinueListOrNext(callResponse, packet, createNextSteps(intialResourceVersion, namespacesToStartNow));
       }
-    },
-    Dedicated {
-      @Override
-      public Collection<String> getDomainNamespaces() {
-        return Collections.singleton(getOperatorNamespace());
+
+      private Step createNextSteps(String intialResourceVersion, Set<String> namespacesToStartNow) {
+        List<Step> nextSteps = new ArrayList<>();
+        if (!namespacesToStartNow.isEmpty()) {
+          nextSteps.add(startNamespaces(Namespaces.this, namespacesToStartNow));
+          if (getConfiguredDomainNamespaces() == null) {
+            nextSteps.add(new InitializeNamespacesSecurityStep(namespacesToStartNow, null));
+          }
+        }
+        nextSteps.add(new CreateNamespaceWatcherStep(selectionStrategy, intialResourceVersion));
+        nextSteps.add(getNext());
+        return Step.chain(nextSteps.toArray(new Step[0]));
       }
-    };
 
-    public boolean isRequireList() {
-      return false;
-    }
+      private Set<String> getNamespacesToStart(List<String> namespaceNames) {
+        return namespaceNames.stream().filter(selectionStrategy::isSelected).collect(Collectors.toSet());
```

Comment:
This is substantially less efficient than the old code as the number of namespaces increases. Why the change? Here's what was done before: 
```
            namespacesToStartNow = namespacesToStartNow.stream().filter(
                    Pattern.compile(regexp).asPredicate()).collect(Collectors.toSet());
```
The new version compiles the regular expression for every namespace tested.

Reply:
It was done for code simplification. I agree that it is less efficient; the real question should be, I think, whether there is a noticeable performance degradation. Do you have a way to test that?

Reply:
I can cache the result, which should achieve roughly the same performance while adding only a small bit of extra code.

Reply:
With the caching, the performance should actually be a bit better, as it should not need to recompile the pattern unless it is changed. I'm still not persuaded that it matters.

Reply:
You could create a test that compiles a slightly more complicated regular expression and measure the time.  The difference in code complexity seems fairly small though between having a method that checks if a namespace matches or returning a predicate function that does the same.

Reply:
The approach I have taken avoids the need for other code to test the strategy; it makes the same call in all cases, and the logic is in the enum. At any rate, the new implementation should be even faster than the original.

Reply:
Let's add a test to demonstrate that the new implementation is still faster as the number of namespaces increases.